### ISSUE TITLE
Data pipeline disappears when there is more than one starting point for the pipeline

### ISSUE THREAD

**Issue by zacsimile** [python-microscopy/python-microscopy#246]
<img width="1012" alt="Screen Shot 2020-05-14 at 1 05 09 PM" src="https://user-images.githubusercontent.com/1263313/81963771-93eb2f00-95e3-11ea-9b94-b92839c6d149.png">

**Comment by zacsimile** [comment]
<img width="1012" alt="Screen Shot 2020-05-14 at 1 11 33 PM" src="https://user-images.githubusercontent.com/1263313/81964355-72d70e00-95e4-11ea-938b-40bfc9d211c2.png">

**Comment by barentine** [comment]
or more than one output
![Screen Shot 2020-05-14 at 1 39 21 PM](https://user-images.githubusercontent.com/31105780/81967027-52a94e00-95e8-11ea-9ed2-9285482254c8.png)

**Comment by David-Baddeley** [comment]
Issue is not present on py2, wx3. 

best guess, it's related to this traceback:
```
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /opt/concourse/worker/volumes/live/b30868b5-2a7c-479c-4e29-27874bbc69a6/volume/wxpython_1547931175613/work/ext/wxWidgets/src/common/wincmn.cpp(478) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.UpdateUIEvent'> returned a result with an error set

```

The implication of which is that it's probably traits not cleaning up after itself properly when we destroy the old recipe display.

Any punters to try on py2, wx4 to confirm this is a wx4 issue?

**Comment by zacsimile** [comment]
Not a wx4 issue. Tried on python 2. Works with wx3 and wx4. Still receive this error with py2/wx4, but the pipeline doesn't disappear.

```
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /opt/concourse/worker/volumes/live/80490b2f-c578-4407-562a-dbbbe365123b/volume/wxpython_1547931284446/work/ext/wxWidgets/src/common/wincmn.cpp(478) in ~wxWindowBase(): any pushed event handlers must have been removed
```

I tried matching the versions of `wxpython`, `traits`, `traitsui`, and `matplotlib` that work on python 2 in the python 3 install. No dice. Also tried downgrading all of `matplotlib`'s dependencies. Exactly the same package sets don't have this error on python 2, but do on python 3.

**Comment by David-Baddeley** [comment]
ugh. I doubt if it's a matplotlib issue (pipeline display on left doesn't use matplotlib). Seeing as it would appear to be py3 rather than wx4 I wonder if it's an iterator vs list issue ...

**Comment by zacsimile** [comment]
Is pipeline display on the left defined in `PYME.recipes.vertical_recipe_display`? I'm not getting any iterator issues, but it could be in a `try` block...

**Comment by David-Baddeley** [comment]
yep - it's vertical recipe display

**Comment by zacsimile** [comment]
Cool. It does use `matplotlib` in two places, but I don't think that's the issue. All the dictionary iterators seem to have been updated. Will keep investigating.

**Comment by zacsimile** [comment]
OK, I think the issue might be in this section at the beginning of `_layout`. 

```
        if self.fp:
             self.fp.elements = []
             self.fp.DestroyChildren()
        self.fp = None
        #print('destroyed fold panel children')
        
        self.DestroyChildren()
        #print('destroyed children')
```

The `self.fp.DestroyChildren()` call yields this traceback on python 3, but no error on python 2.

```
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/buildbots/macosx-vm6/dist-osx-py36/Phoenix/ext/wxWidgets/src/common/wincmn.cpp(478) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/buildbots/macosx-vm6/dist-osx-py36/Phoenix/ext/wxWidgets/src/common/wincmn.cpp(478) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
```

The crazy thing is, for each `foldingPane` in the `self.fp` `foldPanel`, the event handler is the `foldingPane` itself. As such, we can't remove the event handler before removing the pane, it happens simultaneously. However, a quick check after the `DestroyChildren()` call reveals that all `foldingPane`s and the `foldPanel` are removed. As such, this error may be a red herring. However, this error is the only difference in the traceback between python 2 and python 3.

**Comment by zacsimile** [comment]
It was a red herring. Fix is in #334. `sorted()` on line 204 of `PYME.recipes.recipeLayout` is sometimes called on a list of objects, e.g. `[<PYME.recipes.localisations.ProcessColour object at 0x1c24c57308>, <PYME.recipes.surface_fitting.DualMarchingCubes object at 0x1c25dcf410>]`. This is no problem on py2, but `sorted()` needs a key to sort on objects in py3.

**Comment by zacsimile** [comment]
Fixed in #334.
